ophyd/v2/core.py

```diff
@@ -403,86 +408,160 @@ def set_callback(self, callback: Optional[ReadingValueCallback[T]]) -> None:
 
-        if datatype is None:
-            self._initial_value = cast(T, None)
-        elif issubclass(datatype, Enum):
-            self._initial_value = cast(T, list(datatype)[0])
-        else:
-            self._initial_value = datatype()
-        self.set_value(self._initial_value)
+    async def connect(self) -> None:
+        self.converter = make_converter(self.datatype)
+        self._initial_value = self.converter.make_initial_value(self.datatype)
+        self._value = self._initial_value
+        self._severity = 0
+        self._reading = self.converter.reading(
+            self._initial_value, time.monotonic(), self._severity
+        )
 
-    async def connect(self):
-        pass
+        await self.put(self._initial_value)
 
     async def put(self, value: Optional[T], wait=True, timeout=None):
         if value is None:
-            self.set_value(self._initial_value)
+            write_value = self._initial_value
         else:
-            self.set_value(value)
+            write_value = self.converter.write_value(value)
+
+        self._value = write_value
+        self._timestamp = time.monotonic()
+        self._reading = await self.get_reading()
+        if self.callback:
+            self.callback(self._reading, self._value)
+
         if wait:
             await asyncio.wait_for(self.put_proceeds.wait(), timeout)
 
-    async def get_descriptor(self) -> Descriptor:
-        try:
-            dtype = primitive_dtypes[type(self._value)]
-            shape = []
-        except KeyError:
-            if isinstance(self._value, Sequence):
-                dtype = "array"
-                shape = [len(self._value)]
-            elif isinstance(self._value, Enum):
-                dtype = "string"
-                shape = []
-                choices = [e.value for e in type(self._value)]
-                # Ignore type until https://github.com/bluesky/event-model/issues/235#issuecomment-1497353265
-                return dict(source=self.source, dtype=dtype, shape=shape, choices=choices)  # type: ignore
-            else:
-                assert False, f"Can't get dtype for {type(self._value)}"
-        return dict(source=self.source, dtype=dtype, shape=shape)
+    def _set_value(self, value: T):
+        """Method to bypass asynchronous logic, designed to only be used in tests."""
+        self._value = value
+        self._timestamp = time.monotonic()
+        self._reading = self.converter.reading(self._value, self._timestamp, self._severity)
 
-    @property
-    def _reading(self) -> Reading:
-        return dict(value=self._value, timestamp=self._timestamp)
+        if self.callback:
+            self.callback(self._reading, self._value)
+
+    async def get_descriptor(self) -> Descriptor:
+        return self.converter.descriptor(self.source, self._value)
 
     async def get_reading(self) -> Reading:
-        return self._reading
+        return self.converter.reading(self._value, self._timestamp, self._severity)
```

Comment:
Why do we stash `self._reading` sometimes but not here? Maybe we should never stash it, and create it on demand?